### PR TITLE
Add a generic hyperparameter scheduler (closes #104)

### LINKED ISSUE #104: Generic hyperparameter scheduler
The Idea is to have a TF scheduler which is optimizer/hyperparameter agnostic. 
As an example, it can be used for gamma, threshold or learning rate of the floating point optimizer in Bop.

### PR DESCRIPTION
**Pull request by arashb**
this is the first version of the generic HP scheduler.

### CODE REVIEW

**Comment by arashb** [comment]
@koenhelwegen @lgeiger this version passes the tests I developed for with TF 1.14+ but before extending it to TF 1.13: 
1. Do you see any possible problem with this approach? 
2. which of the two interfaces makes more sense? to pass the hyperparameter's name as a string or pass the hyperparaters attribue instead of its name?

File: larq/optimizers_v1.py [hunk 1]
```diff
@@ -181,8 +181,8 @@ def __getattr__(self, name):
     def get_config(self):
         fp_optimizer_config = self.fp_optimizer.get_config()
         config = {
-            "threshold": self.threshold,
-            "gamma": self.gamma,
+            "threshold": tf.keras.backend.get_value(self.threshold),
+            "gamma": tf.keras.backend.get_value(self.gamma),
```
Comment (lgeiger) [inline larq/optimizers_v1.py]: Just to be safe, I think we should cast it to a `float` here to be inline with what [keras does](https://github.com/keras-team/keras/blob/master/keras/optimizers.py#L213):
```suggestion
            "gamma": float(tf.keras.backend.get_value(self.gamma)),
```

File: larq/optimizers_v1.py [hunk 2]
```diff
@@ -181,8 +181,8 @@ def __getattr__(self, name):
     def get_config(self):
         fp_optimizer_config = self.fp_optimizer.get_config()
         config = {
-            "threshold": self.threshold,
-            "gamma": self.gamma,
+            "threshold": tf.keras.backend.get_value(self.threshold),
```
Comment (lgeiger) [inline larq/optimizers_v1.py]:
```suggestion
            "threshold": float(tf.keras.backend.get_value(self.threshold)),
```

File: larq/callbacks_test.py [hunk 3]
```diff
@@ -46,3 +52,58 @@ def test_quantization_logger(self):
...
+            x_train.shape[1], 20, y_train.shape[1]
+        )
+        bop_optimizer = lq.optimizers.Bop(fp_optimizer=tf.keras.optimizers.Adam(0.01))
+        model.compile(
+            loss="categorical_crossentropy",
+            optimizer=bop_optimizer,
+            metrics=["accuracy"],
+        )
+
+        scheduler = lambda x: 1.0 / (1.0 + x)
```
Comment (lgeiger) [inline larq/callbacks_test.py]: I think assigning lambda functions to variables is generally discurraged in Python. Let's just go with a normal function:
```suggestion
        def scheduler(epoch):
            return 1 / (1 + epoch)
```

File: larq/callbacks.py [hunk 4]
```diff
@@ -74,3 +74,46 @@ def on_train_begin(self, logs=None):
...
+            hyperparameter_val = self.schedule(epoch, hyperparameter_val)
+        except TypeError:  # Support for old API for backward compatibility
+            hyperparameter_val = self.schedule(epoch)
+
+        tf.keras.backend.set_value(hp, hyperparameter_val)
+
+        if self.verbose > 0:
+            print(
+                "\nEpoch %05d: %s changning to %s."
+                % (epoch + 1, self.hyperparameter, tf.keras.backend.get_value(hp))
```
Comment (lgeiger) [inline larq/callbacks.py]: I think we can reuse `hyperparameter_val` here, correct?
```suggestion
                % (epoch + 1, self.hyperparameter, hyperparameter_val)
```

File: larq/callbacks.py [hunk 5]
```diff
@@ -74,3 +74,46 @@ def on_train_begin(self, logs=None):
...
+            hyperparameter_val = tf.keras.backend.get_value(hp)
+            hyperparameter_val = self.schedule(epoch, hyperparameter_val)
+        except TypeError:  # Support for old API for backward compatibility
+            hyperparameter_val = self.schedule(epoch)
+
+        tf.keras.backend.set_value(hp, hyperparameter_val)
+
+        if self.verbose > 0:
+            print(
+                "\nEpoch %05d: %s changning to %s."
```
Comment (lgeiger) [inline larq/callbacks.py]: I really like the new f-strings in Python which make this syntax a lot more readable:
```python
print(f"Epoch {epoch + 1}: {self.hyperparameter} changing to {hyperparameter_val}")
```

File: larq/callbacks.py (hunk 4 above)
Comment (arashb) [reply larq/callbacks.py]: yes, but I wanted to print out the value that is actually stored in optimizer. This is helpful in cases where implicit casting is happening.

Comment (lgeiger) [reply larq/callbacks.py]: I see. For casting it makes sense.

**Pull request merged**